### ISSUE TITLE
TCA somehow interfering with a view by anonymous users

### ISSUE THREAD

**Issue by nattywebdev** [backdrop-contrib/tca#3]
Delighted to see TCA ported to Backdrop - thank you!

I have a weird thing happening when the TCA module is enabled, but not activated for (eg) content type 'Page'. On my site, pages have a View that shows a banner image created as a Paragraph for that page. This works fine for my admin account but when viewed as Anonymous, the View (banner image) is not displayed. If I disable the TCA module, all works fine. Re-enable it and the View is hidden again.

I'd really like to use the TCA module as it fits perfectly a particular use case. Can you suggest what might be going on here? Or where to look to find further clues?

**Comment by yorkshire-pudding** [comment]
I'm a bit confused by the combination of View and Paragraph. Please can you provide more details of how this is defined. I'd like to try and reproduce so I can see if this is a bug, or just needs guidance.

Detailed steps to reproduce would be helpful. If this is a publicly available site (even if not live), please can you share the URL, privately on Zulip would be fine.

**Comment by nattywebdev** [comment]
It's a single Paragraph of type 'banner', which comprises image, page title and subtitle. Originally it was built with a Field Collection, but I changed it for this site to use Paragraphs as that was more consistent (and meant I could remove Field Collections). The problem was there when the node type was using Field Collections, and I switched to Paragraphs as that was my intention. As you might guess, this is a D7 upgrade.
 I'll share the site details in Zulip.

The View is used to display the single Paragraph (banner) at the region called 'Top' at the top of the page, if one exists; for Layouts I'm using Harris (flexible).

**Comment by nattywebdev** [comment]
Reports > Recent Log messages shows this line, which might be pertinent:
Exception in Banner paragraph[banner_paragraph]: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'node.tca_active' in 'field list'

**Comment by yorkshire-pudding** [comment]
I'm going to try and reproduce this.

**Comment by yorkshire-pudding** [comment]
Point of clarification for reproducing this: How is the paragraph banner item disabled from view as part of the paragraphs field?

**Comment by nattywebdev** [comment]
Like so:
![image](https://github.com/user-attachments/assets/43ad190e-784b-4bb3-9d1a-437ac6a90c6b)

**Comment by nattywebdev** [comment]
Doesn't show there but the 'Banner' is a Paragraphs field with just one Paragraph Type enabled (named as 'Banner').

**Comment by yorkshire-pudding** [comment]
I can't reproduce this with the information given here and in Zulip.
My steps:

1. create content type with paragraphs field and other field (i.e. body) - Don't enable TCA for this content type
2. Hide the para field from display
3. create "banner" para type with image and caption
4. create test node with a banner and content
5. create view block
6. Add block to layout
7. View page

I wonder if something else is happening here

<details><summary>Images</summary>

**View**
![image](https://github.com/user-attachments/assets/00b343a4-5823-4173-9a28-e9951a71e0f9)


**Logged in**
![image](https://github.com/user-attachments/assets/e31db482-a4f4-4a66-a661-d50c7631ee20)

**Anon**
![image](https://github.com/user-attachments/assets/7f40bea7-4b0b-40ad-9412-b39f1f26dc72)

</details>

**Comment by nattywebdev** [comment]
Yes, I suspect it might be. Another inherited site upgraded from D7. Frustrating!

**Comment by yorkshire-pudding** [comment]
Hi @nattywebdev - Just wondering if you made any progress in narrowing this down?

**Comment by nattywebdev** [comment]
Nope. The website development came to a grinding halt - not because of this but other pressures on the customer's time - and so I didn't pursue it any further. I simply couldn't get it to work using that particular config (banner shown using paragraphs in a view).

If the work resumes, I'll probably do things a different way. Shame, as the TCA module was potentially perfect.

Thanks for asking though!

**Comment by laryn** [comment]
May have been access/permissions on the view in question.

**Comment by yorkshire-pudding** [comment]
@laryn - @nattywebdev shared some site details privately that appeared to show that the view were set with the correct (View published content) permission. I don't think sharing those screenshots here will be a problem as there is no confidential information in them:

> The content for the view comes from the Paragraphs item: Bundle = Page banner, as you can see here:
> ![Image](https://github.com/user-attachments/assets/13e8285f-771a-4113-94ac-f4970d6ff5dc)
> 
> The default page display is set to hide that particular paragraph type, so that it can be displayed in a different region ("Top") on the relevant page.
> 
> The content types where this View is displayed (Front Page and Page) both have the TCA option set to OFF. I only want to use it on one particular content type, where it will be set to ON.
> 
> I'm worried in case I'm not using TCA correctly...
> 
> I wondered if it was the back-to-front way I'd constructed the View, so changed it to show banner paragraphs for the current node:
> 
> ![Image](https://github.com/user-attachments/assets/af722c7c-240d-48fb-ae64-512655e174e7)
> 
> Result? No difference! So I'm baffled and may simply have to drop use of TCA (unless you spot where I'm going wrong).
> 
> weird thing is it works fine (for Anon) on my local Lando dev site. I've checked everything is the same (it is, including php version). Only difference I know is the staging site is using nginx but no idea if that could affect anything.

It would be good to get to the bottom of it, because I can't figure out how TCA could cause this just by being enabled on the site.

**Comment by argiepiano** [comment]
>  This works fine for my admin account but when viewed as Anonymous, the View (banner image) is not displayed

Hmmm.... this sounds like a cached page. @nattywebdev have you tried going to `admin/config/development/performance`  and disabling the cache for anonymous users, then test again to see if the issue persists?

**Comment by laryn** [comment]
Possibly connected to https://github.com/backdrop/backdrop-issues/issues/6421

**Comment by argiepiano** [comment]
> Possibly connected to https://github.com/backdrop/backdrop-issues/issues/6421

Hmmm that's farfetched. That bug only applies when Content Access (or other modules that implement `hook_node_grants()`) are enabled. Otherwise that bug did not affect any joins or views at all.

**Comment by laryn** [comment]
Okay, I retract that then. :) I know nothing about this module except that it's access-related and was having trouble with a view not displaying content that should be displayed, which made the other issue seem plausible. If you know more about this module and think that's not the case I trust your opinion.

**Comment by nattywebdev** [comment]
I've revisited this problem as detailed in my first post. It still causes the same problem for me, that is:

- if I tick 'Enable Token Content Access protection' for a content type (eg: Page) and then enable that for a particular page, it works as I expect it to: the page is only accessible to anonymous user if I access it using the provided token.
- however, the View normally shown at the top of that page (showing the page title or a banner image if available) does not display.

In short: the problem persists for me. I am not using Content Access at all for this site.

@argiepiano  There is no caching on (`admin/config/development/performance`) while I'm developing the site, so it's not a caching issue.

**Comment by yorkshire-pudding** [comment]
@nattywebdev - I have another idea that may be an actual bug with the module.
Is it possible that Token Content Access was once enabled on this content type and activated for this node?  You can check this by either:

- checking to see if in the `node` database tables those nodes have a `1` rather than `0` or `NULL` in column `tca_active`
- enabling for the node type and then checking to see if TCA is active on the node.

I have discovered in testing, that if the token is set for a node and then TCA is disabled for the content type then it is still applying the protection which is not what it is meant to do; I will tackle this bug whether this is the cause here, or I'll raise separately if not.

**Comment by argiepiano** [comment]
> I know nothing about this module except that it's access-related and was having trouble with a view not displaying content that should be displayed, which made the other issue seem plausible. 

Thanks, @laryn - it IS a good lead. I have to admit I don't know anything about this module either :)

**Comment by argiepiano** [comment]
@nattywebdev, following up on @laryn's lead, perhaps you are using a module that implements hook_node_grants, in which case @laryn's assessment may be csorrect. It may be a good idea to go to debug at `admin/reports/debug` and paste here the list of contrib modules you are using.

**Comment by nattywebdev** [comment]
@yorkshire-pudding although the token is still there in the node table (as tca_token) for any nodes I've ever added TCA for, tca_active is correctly set as 0 where disabled. 

So no, doesn't look like that's the cause.

List of modules follows

```
Enabled modules
===============

Core
----
admin_bar                            1.28.2
block                                1.28.2
ckeditor                             1.28.2
ckeditor5                            1.28.2
config                               1.28.2
contextual                           1.28.2
date                                 1.28.2
dblog                                1.28.2
entity                               1.28.2
entityreference                      1.28.2
field                                1.28.2
field_sql_storage                    1.28.2
field_ui                             1.28.2
file                                 1.28.2
filter                               1.28.2
image                                1.28.2
installer                            1.28.2
layout                               1.28.2
link                                 1.28.2
list                                 1.28.2
menu                                 1.28.2
node                                 1.28.2
options                              1.28.2
path                                 1.28.2
redirect                             1.28.2
search                               1.28.2
system                               1.28.2
taxonomy                             1.28.2
text                                 1.28.2
update                               1.28.2
user                                 1.28.2
views                                1.28.2
views_ui                             1.28.2

Contrib
-------
admin_menu_source                    1.x-1.0.1
backup_migrate                       1.x-1.0.26
better_exposed_filters               1.x-3.6.2
better_search                        1.x-1.1.1
chosen                               1.x-2.1.2
classy_paragraphs                    1.x-1.1.5
classy_paragraphs_ui                 1.x-1.1.5
colectomy                            1.x-1.0.2
colorbox                             1.x-2.17.2
draggableviews                       1.x-1.0.1
editorimgdimensionsync               1.x-1.0.1
entity_plus                          1.x-1.0.21
eu_cookie_compliance                 1.x-2.26.0
extlink                              1.x-1.18.2
feeds                                1.x-1.1.5
feeds_tamper                         1.x-1.2.3
feeds_tamper_ui                      1.x-1.2.3
feeds_ui                             1.x-1.1.5
field_group                          1.x-1.2.1
file_resup                           1.x-1.2.0
filefield_paths                      1.x-1.0.4
image_resize_filter                  1.x-1.16.3
job_scheduler                        1.x-1.1.4
masquerade                           1.x-1.0.0
menu_expanded                        1.x-1.0.0
metatag                              1.x-1.22.6
paragraphs                           1.x-1.3.0
preset                               1.x-2.0.2
schema_metatag                       1.x-1.0.2-beta
schema_recipe                        1.x-1.0.2-beta
schema_web_site                      1.x-1.0.2-beta
smart_trim                           1.x-1.1.3
smtp                                 1.x-1.7.1
special_menu_items                   1.x-2.1.3
stickynav                            1.x-1.0.4
tca                                  1.x-1.1.0
tca_node                             1.x-1.1.0
views_aggregator                     1.x-1.0.6
views_aggregator_more_functions      1.x-1.0.6
views_data_export                    1.x-3.2.0
views_filters_selective              1.x-1.0.0

Other
-----
field_collection                     -
googleanalytics                      -

```

**Comment by nattywebdev** [comment]
Interesting there is no version for field_collection or googleanalytics. I'll disable and uninstall both; the former I've been removing slowly, that latter is no use until the site is live.

No disabled, uninstalled, cache cleared. And the problem remains!

**Comment by yorkshire-pudding** [comment]
Not sure if it will help for this, but I've documented the issue I found and added a PR at #7 - might be worth testing it against this issue.

**Comment by yorkshire-pudding** [comment]
@nattywebdev - not sure it will fix this issue, but worth updating to the latest version (with the fix mentioned above) to see if it does make a difference to this.

**Comment by nattywebdev** [comment]
I downloaded the latest version and it made no difference to this issue - but thanks for checking!

**Comment by nattywebdev** [comment]
My example was more complex than it needed to be (with a banner image with paragraphs). The same problem occurs when I have a page whose Title is displayed by a view in a block at the top of the page.

**Comment by argiepiano** [comment]
@nattywebdev I'm curious about this issue (sorry, I'm a glutton for punishment). Can you please describe step by step how to reproduce it? While this has been described in several postings above, it'd be easier to follow if the steps are clearly laid out in one posting. Maintainers, I hope you don't mind my "meddling" here 😄

**Comment by yorkshire-pudding** [comment]
Feel free to meddle @argiepiano 

I just attempted with a clean install of all the modules and still cannot reproduce.  I suspect there is some config or permissions we are missing.

**Comment by yorkshire-pudding** [comment]
This is a longshot @nattywebdev but I note you have Backup and Migrate.  Would you be willing to **privately** share a full backup so we can reproduce and debug?
I appreciate that this may not be permissible

**Comment by nattywebdev** [comment]
I don't see why not, in the interests of trying to make a feature work. I'll message you on Zulip

**Comment by nattywebdev** [comment]
@argiepiano happy to:

1. install & enable the TCA module;
2. for the 'Page' content type, tick to Enable Token Content Access protection;
3. for an existing Page, tick to Enable Token Content Access protection;
4. check the page is accessible with the new token; check it isn't without the token;
5. in Layouts (default layout) configure the Page title to show none (in my layout this is between the Header and Top sections);
6. check the page (above) shows but without a title;
7. add a view (block) that shows the title of the current node (contextual filter: nid from content id from url);
8. display this block in Layouts (default) in the Top region;
9. check this works on a page where the TCA hasn't been ticked;
10. display the page where the TCA has been ticked

The result for me is the Title view is not displayed on the page where the TCA has been ticked, for anonymous users.

When the TCA is unticked, the Title view is displayed again.

**Comment by yorkshire-pudding** [comment]
@nattywebdev - a slight misunderstanding.  I thought from:

> I have a weird thing happening **when the TCA module is enabled, but not activated for (eg) content type 'Page'.** On my site, pages have a View that shows a banner image created as a Paragraph for that page. This works fine for my admin account but when viewed as Anonymous, the View (banner image) is not displayed. If I disable the TCA module, all works fine. Re-enable it and the View is hidden again.

That the content type page was not activated.

If the content type and the page are activated for TCA then all you need to do in the view is:

![Image](https://github.com/user-attachments/assets/530f90b1-e17a-4ad4-a33e-fb37d217fe5a)

That will expose to people without the token too but they won't see the rest of the page. 

Can you check that please?

**Comment by yorkshire-pudding** [comment]
If you want to also hide from people without the token there is a way to do that using a contextual filter but its a bit more complex

**Comment by nattywebdev** [comment]
That does indeed do the trick. Not sure I'd ever have noticed that! Is it mentioned somewhere in documentation? If so, I am sincerely sorry for wasting everyone's time!

I would like to be able to hide the view from people without the token - can you give me a steer on that please @yorkshire-pudding ?

**Comment by yorkshire-pudding** [comment]
> I would like to be able to hide the view from people without the token - can you give me a steer on that please [@yorkshire-pudding](https://github.com/yorkshire-pudding) ?

I will try to but the scant documentation I had made for myself is not working as I remember it doing when I first added the view functionality

**Comment by yorkshire-pudding** [comment]
Here are some brief instructions.  I'll look to do a Wiki page I think with some screenshots, but this should be enough to get you going.

In view display:

1. Other > Token Content Access Node Override: Yes
2. Contextual Filters > Add > Token Content Access Token - select
3. Add and Configure Contextual Filters
4. When the filter value is NOT available > Provide default value > Type > PHP Code
```php
if (isset($_GET['tca'])) {
  return $_GET['tca'];
}
elseif (user_access('bypass tca_node')) {
    $node = menu_get_object();
    return $node->tca_token;
  }
else {
  return;
}
```
5.  Skip default argument for view URL = unchecked
6. When the filter value IS available or a default is provided > Specify validation criteria = checked
7. Validator > '- Basic validation - '
8. Action to take if filter value does not validate > Hide view

If you want to preview in the view config, add the node id and the token (e.g. `2/2A2VLheZynTG5SJ0DYas8zzcGBXPciM_i0eh1iKOdvY`) to 'Preview with contextual filters.

**Comment by nattywebdev** [comment]
@yorkshire-pudding - thank you - that works exactly as expected.
